src/Adapter/Discount/Validate/DiscountValidator.php

```diff
@@ -81,6 +83,50 @@ public function validate(CartRule $cartRule): void
         $this->assertCodeIsUnique($cartRule);
     }
 
+    /**
+     * @throws DiscountConstraintException
+     */
+    public function validateDiscountPropertiesForType(AddDiscountCommand $command)
+    {
+        switch ($command->getDiscountType()->getValue()) {
+            case DiscountType::FREE_SHIPPING:
+                if ($command->isFreeShipping()) {
+                    break;
+                } else {
+                    throw new DiscountConstraintException('A free shipping discount must have the free shipping property set to true.', DiscountConstraintException::INVALID_FREE_SHIPPING);
+                }
+                // no break
+            case DiscountType::CART_DISCOUNT:
+            case DiscountType::ORDER_DISCOUNT:
+                if ($command->getAmountDiscount() !== null && $command->getPercentDiscount() !== null) {
+                    throw new DiscountConstraintException('Discount can not be amount and percent at the same time', DiscountConstraintException::INVALID_DISCOUNT_CANNOT_BE_AMOUNT_AND_PERCENT);
+                }
+                if ($command->getAmountDiscount() !== null) {
+                    if ($command->getAmountDiscount()->getAmount()->isLowerThanZero()) {
+                        throw new DiscountConstraintException('Discount value can not be negative', DiscountConstraintException::INVALID_DISCOUNT_VALUE_CANNOT_BE_NEGATIVE);
+                    }
+                }
+                if ($command->getPercentDiscount() !== null) {
+                    if ($command->getPercentDiscount()->isLowerThanZero()) {
+                        throw new DiscountConstraintException('Discount value can not be negative', DiscountConstraintException::INVALID_DISCOUNT_VALUE_CANNOT_BE_NEGATIVE);
+                    }
+                }
+                break;
+            case DiscountType::PRODUCTS_DISCOUNT:
+                if ($command->getReductionProduct() === 0 || $command->getPercentDiscount() === null) {
+                    throw new DiscountConstraintException('Product discount must have his properties set.', DiscountConstraintException::INVALID_PRODUCT_DISCOUNT_PROPERTIES);
+                }
+                break;
+            case DiscountType::FREE_GIFT:
+                if ($command->getProductId() === null) {
+                    throw new DiscountConstraintException('Free gift discount must have his properties set.', DiscountConstraintException::INVALID_FREE_GIFT_DISCOUNT_PROPERTIES);
+                }
+                break;
+            default:
+                throw new DiscountConstraintException(sprintf("Invalid discount type '%s'.", $command->getDiscountType()->getValue()), DiscountConstraintException::INVALID_DISCOUNT_TYPE);
```

Comment:
Good initiative for now But we may need to remove this later when we'll deal with the extensibility of the feature For example if someone wants to create their custom discount type
But it's too early to think about this in details